apio-consumer/src/main/kotlin/com/liferay/apio/consumer/parser/ThingParser.kt

```diff
@@ -0,0 +1,216 @@
+        }
+
+        private fun Embedded<*, *>.parse(context: Context?, things: MutableMap<String, Thing?>,
+            attributes: MutableMap<String, Any>, key: String) {
+
+            (this as? Embedded<String, Any>)?.apply {
+                if (this.containsKey("@id")) {
+                    val result = flatten(this, context)
+
+                    result?.let {
+                        val (thing, embeddedThings) = it
+                        attributes[key] = Relation(thing.id)
+
+                        things.put(thing.id, thing)
+
+                        things.putAll(embeddedThings)
+                    }
+                } else {
+                    val (attr, embeddedThings) = foldTree(this, context)
+
+                    things.putAll(embeddedThings)
+                    attributes[key] = attr
+                }
+            }
+        }
+
+        private fun EmbeddedList<*>.parse(context: Context?, things: MutableMap<String, Thing?>,
+            attributes: MutableMap<String, Any>, key: String) {
+
+            (this as? EmbeddedList<Embedded<String, Any>>)?.apply {
+                if (this.first().containsKey("@id")) {
+                    val list = this.map { flatten(it, context)!! }
+                    val mutableList = mutableListOf<Relation>()
+
+                    for ((thing, embeddedThings) in list) {
+                        mutableList.add(Relation(thing.id))
+
+                        things.put(thing.id, thing)
+
+                        things.putAll(embeddedThings)
+                    }
+
+                    attributes[key] = mutableList
+                } else {
+                    val pairsList = this.map { foldTree(it, context) }
+
+                    val foldedList = mutableListOf<Map<String, Any>>()
+
+                    for ((list, embeddedThings) in pairsList) {
+                        foldedList.add(list)
+
+                        things.putAll(embeddedThings)
+                    }
+
+                    attributes[key] = foldedList
+                }
+            }
+        }
+
+        private fun EmbeddedList<*>.hasMapInstances() = !this.filterIsInstance<Map<String, Any>>().isEmpty()
+
+        private fun parseOperations(jsonObject: Map<String, Any>): MutableMap<String, Operation> {
+            val operationsJson = jsonObject["operation"] as? List<Map<String, Any>> ?: listOf()
+
+            return operationsJson.map {
+                val id = it["@id"] as String
+                val target = it["target"] as String
+                val method = it["method"] as String
+                val expects = it["expects"] as? String
+                val type = parseType(it["@type"])
+
+                val form = expects?.let { OperationForm(it) }
+
+                id to Operation(id, target, type, method, form)
+            }.toMap().toMutableMap()
+        }
+    }
+
+}
+
+typealias Embedded<T, U> = Map<T, U>
+typealias EmbeddedList<T> = List<T>
```

Comment:
👍

Reply:
Just readability, we were trying to understand the parser, so we did that. The meaning is correct? @victorg1991

Reply:
Yep :)